tests/e2e/websocket/websocket.cpp

```diff
@@ -0,0 +1,214 @@
+  }
+
+  void OnResolve(beast::error_code ec, tcp::resolver::results_type results) {
+    if (ec) {
+      return Fail(ec, "resolve");
+    }
+
+    beast::get_lowest_layer(ws_).expires_after(std::chrono::seconds(30));
+    beast::get_lowest_layer(ws_).async_connect(results,
+                                               beast::bind_front_handler(&Session::OnConnect, shared_from_this()));
+  }
+
+  void OnConnect(beast::error_code ec, tcp::resolver::results_type::endpoint_type ep) {
+    if (ec) {
+      return Fail(ec, "connect");
+    }
+
+    beast::get_lowest_layer(ws_).expires_never();
+
+    ws_.set_option(websocket::stream_base::timeout::suggested(beast::role_type::client));
+
+    ws_.set_option(websocket::stream_base::decorator([](websocket::request_type &req) {
+      req.set(http::field::user_agent, std::string(BOOST_BEAST_VERSION_STRING) + " websocket-client-async");
+    }));
+
+    host_ = fmt::format("{}:{}", host_, ep.port());
+
+    ws_.async_handshake(host_, "/", beast::bind_front_handler(&Session::OnHandshake, shared_from_this()));
+  }
+
+  void OnHandshake(beast::error_code ec) {
+    if (ec) {
+      return Fail(ec, "handshake");
+    }
+    ws_.async_write(net::buffer(GetAuthenticationJSON(creds_)),
+                    beast::bind_front_handler(&Session::OnWrite, shared_from_this()));
+  }
+
+  void OnWrite(beast::error_code ec, std::size_t bytes_transferred) {
+    boost::ignore_unused(bytes_transferred);
+
+    if (ec) {
+      return Fail(ec, "write");
+    }
+
+    ws_.async_read(buffer_, beast::bind_front_handler(&Session::OnRead, shared_from_this()));
+  }
+
+  void OnRead(beast::error_code ec, std::size_t /*bytes_transferred*/) {
+    if (ec) {
+      return Fail(ec, "read");
+    }
+    received_messages_.push_back(boost::beast::buffers_to_string(buffer_.data()));
+    buffer_.clear();
+
+    ws_.async_read(buffer_, beast::bind_front_handler(&Session::OnRead, shared_from_this()));
+  }
+
+  void OnClose(beast::error_code ec) {
+    if (ec) {
+      return Fail(ec, "close");
+    }
+  }
+
+ private:
+  tcp::resolver resolver_;
+  websocket::stream<beast::tcp_stream> ws_;
+  beast::flat_buffer buffer_;
+  std::string host_;
+  std::vector<std::string> &received_messages_;
+  Credentials creds_;
+};
+
+class WebsocketClient {
+ public:
+  WebsocketClient() : session_{std::make_shared<Session>(ioc_, received_messages_)} {}
+
+  WebsocketClient(Credentials creds) : session_{std::make_shared<Session>(ioc_, received_messages_, creds)} {}
+
+  void Connect(const std::string host, const std::string port) {
+    session_->Run(host, port);
+    bg_thread_ = std::thread([this]() { ioc_.run(); });
+    bg_thread_.detach();
+  }
+
+  void Close() { ioc_.stop(); }
+
+  std::vector<std::string> GetReceivedMessages() { return received_messages_; }
+
+ private:
+  std::vector<std::string> received_messages_{};
+  net::io_context ioc_{};
+  std::thread bg_thread_;
```

Comment:
```suggestion
  std::jthread bg_thread_;
```

Just to make sure we join the thread on destruction.

Reply:
Why didn't you use `jthread` here but you use it in the SSL version.

Reply:
I would leave the `jthread` so it's closed correctly if `AwaitClose` wasn't called (safer).
Also, `AwaitClose` will allow you to stop the processing manually without relying on RAII so it doesn't lose its functionality.